From 876a00129c58d9b7b161c613fa6e07fb0c2402fd Mon Sep 17 00:00:00 2001
From: pytorchbot <soumith+bot@pytorch.org>
Date: Fri, 17 Jan 2025 17:02:28 -0600
Subject: [PATCH] Prevent legacy_load when weights_only=True (correctly)
 (#145111)

Prevent legacy_load when weights_only=True (correctly) (#145020)

Only prevent `legacy_load` (.tar format removed in https://github.com/pytorch/pytorch/pull/713), not the whole of `_legacy_load` (.tar format + _use_new_zipfile_serialization=False)

Differential Revision: [D68301405](https://our.internmc.facebook.com/intern/diff/D68301405)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/145020
Approved by: https://github.com/kit1980, https://github.com/albanD

(cherry picked from commit 0eda02a94c754e2256ff1701bcc03c40ece2bbef)

[dannf: backported to pytorch 2.3.1]
Co-authored-by: Mikayla Gawarecki <mikaylagawarecki@gmail.com>
---
 test/test_serialization.py |  6 +++++-
 torch/serialization.py     | 16 +++++++++++-----
 2 files changed, 16 insertions(+), 6 deletions(-)

diff --git a/test/test_serialization.py b/test/test_serialization.py
index 5e8a4f0a6e5..85993a9cbf3 100644
--- a/test/test_serialization.py
+++ b/test/test_serialization.py
@@ -425,7 +425,11 @@ class SerializationMixin:
         b += [a[0].storage()]
         b += [a[0].reshape(-1)[1:4].clone().storage()]
         path = download_file('https://download.pytorch.org/test_data/legacy_serialized.pt')
-        c = torch.load(path, weights_only=weights_only)
+        if weights_only:
+            with self.assertRaisesRegex(RuntimeError,
+                                        "Cannot use ``weights_only=True`` with files saved in the legacy .tar format."):
+                c = torch.load(path, weights_only=weights_only)
+        c = torch.load(path, weights_only=False)
         self.assertEqual(b, c, atol=0, rtol=0)
         self.assertTrue(isinstance(c[0], torch.FloatTensor))
         self.assertTrue(isinstance(c[1], torch.FloatTensor))
diff --git a/torch/serialization.py b/torch/serialization.py
index b344a8aade1..99b98155e03 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -885,6 +885,11 @@ def _save(obj, zip_file, pickle_module, pickle_protocol, _disable_byteorder_reco
         num_bytes = storage.nbytes()
         zip_file.write_record(name, storage, num_bytes)
 
+UNSAFE_MESSAGE = (
+   "Weights only load failed. Re-running `torch.load` with `weights_only` set to `False`"
+   " will likely succeed, but it can result in arbitrary code execution."
+    "Do it only if you get the file from a trusted source. WeightsUnpickler error: "
+)
 
 def load(
     f: FILE_LIKE,
@@ -994,11 +999,6 @@ def load(
         >>> torch.load('module.pt', encoding='ascii', weights_only=False)
     """
     torch._C._log_api_usage_once("torch.load")
-    UNSAFE_MESSAGE = (
-        "Re-running `torch.load` with `weights_only` set to `False` will likely succeed, "
-        "but it can result in arbitrary code execution. Do it only if you got the file from a "
-        "trusted source."
-    )
     DOCS_MESSAGE = (
         "\n\nCheck the documentation of torch.load to learn more about types accepted by default with "
         "weights_only https://pytorch.org/docs/stable/generated/torch.load.html."
@@ -1202,6 +1202,11 @@ def _legacy_load(f, map_location, pickle_module, **pickle_load_args):
         with closing(tarfile.open(fileobj=f, mode='r:', format=tarfile.PAX_FORMAT)) as tar, \
                 mkdtemp() as tmpdir:
 
+            if pickle_module is _weights_only_unpickler:
+                raise RuntimeError(
+                    "Cannot use ``weights_only=True`` with files saved in the "
+                    "legacy .tar format. " + UNSAFE_MESSAGE
+                )
             tar.extract('storages', path=tmpdir)
             with open(os.path.join(tmpdir, 'storages'), 'rb', 0) as f:
                 num_storages = pickle_module.load(f, **pickle_load_args)

